code/network/lib/Loot/Network/Class.hs

```diff
@@ -0,0 +1,176 @@
+type Content = [ByteString]
+
+-- | Bytestring describing the type of the subscription. The key.
+newtype Subscription = Subscription { unSubscription :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Message type is characterized as a bytestring.
+newtype MsgType = MsgType { unMsgType :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Two-ended queue, first parameter is messages user receives,
+-- second -- messages user sends.
+data BiTQueue r s = BiTQueue
+    { bReceiveQ :: TQueue r
+      -- ^ Queue to receive messages from.
+    , bSendQ    :: TQueue s
+      -- ^ Queue to send messages to.
+    }
+
+----------------------------------------------------------------------------
+-- Client
+----------------------------------------------------------------------------
+
+-- | Request to update peers
+data UpdatePeersReq nId = UpdatePeersReq
+    { _uprAdd :: Set nId -- ^ Peers to add/connect.
+    , _uprDel :: Set nId -- ^ Peers to delete/disconnect.
+    } deriving (Show, Generic)
+
+makeLenses ''UpdatePeersReq
+
+instance (Ord s) => Default (UpdatePeersReq s) where
+    def = UpdatePeersReq mempty mempty
+
+-- | Either a response from some server or a subscription update (key
+-- and data).
+data CliRecvMsg
+    = Response MsgType Content
+    | Update Subscription Content
+    deriving (Eq,Ord,Show)
+
+-- | Client worker environment.
+type ClientEnv t = BiTQueue (NodeId t, CliRecvMsg) (Maybe (NodeId t), (MsgType, Content))
+
+-- | Client worker identifier.
+type ClientId = ByteString
+
+-- | Client-side networking interface.
+class (Monad m, Ord (NodeId t)) => NetworkingCli t m where
+    -- | Full-size identities -- ports/hosts. Something we can connect to.
+    type NodeId t
+
+    -- | Runs client routine. This must be run in the same thread
+    -- client context was created.
+    runClient :: m ()
+
+    -- | Returns the list of current peers connected.
+    getPeers :: m (Set (NodeId t))
+
+    -- | First arguments -- peers to connect to, second -- to disconnect.
+    updatePeers :: UpdatePeersReq (NodeId t) -> m ()
+
+    -- | Register client worker. Client identifier, msgtype set should
+    -- be unique for a client (to implement proper
+    -- routing). Subscription sets of clients can intersect (several
+    -- clients can subscribe to the same key)..
+    registerClient :: ClientId -> Set MsgType -> Set Subscription -> m (ClientEnv t)
+
+----------------------------------------------------------------------------
+-- Server
+----------------------------------------------------------------------------
+
+-- | Things server sends -- either replies (to the requested node) or
+-- publishing content.
+data ServSendMsg cliId
+    = Reply cliId MsgType Content
+    | Publish Subscription Content
+    deriving (Eq,Ord,Show)
+
+-- | Listener environment.
+type ListenerEnv t = BiTQueue (CliId t, MsgType, Content) (ServSendMsg (CliId t))
+
+-- | Listener identifier.
+type ListenerId = ByteString
```

Comment:
Are the `ClientId`s and `ListenerId`s interchangeable from domain standpoint?

Reply:
Obscure logs are not fine; however lets hope that will not ruin someone's night.